instance/models/mixins/storage.py

```diff
@@ -307,53 +307,68 @@ def _create_bucket(self, max_tries=4, retry_delay=4, location=None):
         location_constraint = location or self.s3_region
         while not bucket:
             try:
-                if not location_constraint or location_constraint == 'us-east-1':
-                    # oddly enough, boto3 uses 'us-east-1' as default and doesn't accept it explicitly
-                    # https://github.com/boto/boto3/issues/125
-                    bucket = self.s3.create_bucket(Bucket=self.s3_bucket_name)
-                else:
-                    bucket = self.s3.create_bucket(
-                        Bucket=self.s3_bucket_name,
-                        CreateBucketConfiguration={
-                            'LocationConstraint': location_constraint
-                        },
-                    )
+                bucket = self._perform_create_bucket(location_constraint)
             except ClientError as e:
                 if e.response.get('Error', {}).get('Code') == 'EntityAlreadyExists':
+                    # Continue if bucket already exists, i.e. reprovisioning
                     self.logger.info(
                         'Bucket %s already exists',
                         self.s3_bucket_name
                     )
                     break
+                # Retry up to `max_tries` times
                 self.logger.info(
-                    'Retrying bucket creation. IAM keys are not propagated yet, attempt %s of %s.',
-                    attempt, max_tries
+                    'Retrying bucket creation due to "%s", attempt %s of %s.',
+                    e.reponse.get('Error', {}).get('Code'), attempt, max_tries
                 )
                 time.sleep(retry_delay)
                 attempt += 1
                 if attempt > max_tries:
                     raise
 
-        # Set bucket cors
-        self.s3.put_bucket_cors(
+        # Update bucket cors
+        self._update_bucket_cors()
+
+        # Update bucket lifecycle configuration
+        self._update_bucket_lifecycle()
+
+        # Enable bucket versioning
+        self._enable_bucket_versioning()
+
+    def _enable_bucket_versioning(self):
+        self.s3.put_bucket_versioning(
             Bucket=self.s3_bucket_name,
-            CORSConfiguration=S3_CORS
+            VersioningConfiguration={
+                'Status': 'Enabled'
+            }
         )
```

Comment:
Do we need to catch any errors here?

Reply:
Those errors are bubbling up to the `_create_bucket` where it's being retried. In case of failure, the redeployment/provisioning must fail.